contrib/node/src/python/pants/contrib/node/tasks/node_resolve.py

```diff
@@ -66,23 +72,58 @@ def _can_resolve_target(self, target):
-
-    # We must have copied local sources into place and have node_modules directories in place for
-    # internal dependencies before installing dependees, so `topological_order=True` is critical.
-    with self.invalidated(targets,
-                          topological_order=True,
-                          invalidate_dependents=True) as invalidation_check:
-
-      with self.context.new_workunit(name='install', labels=[WorkUnitLabel.MULTITOOL]):
+    if self.context.products.is_required_data(NodePaths):
+      node_paths = self.context.products.get_data(NodePaths, init_func=NodePaths)
+      # We must have copied local sources into place and have node_modules directories in place for
+      # internal dependencies before installing dependees, so `topological_order=True` is critical.
+      with self.invalidated(targets,
+                            topological_order=True,
+                            invalidate_dependents=True) as invalidation_check:
+        with self.context.new_workunit(name='install', labels=[WorkUnitLabel.MULTITOOL]):
+          for vt in invalidation_check.all_vts:
+            target = vt.target
+            if not vt.valid:
+              resolver_for_target_type = self._resolver_for_target(target).global_instance()
+              resolver_for_target_type.resolve_target(self, target, vt.results_dir, node_paths)
+            node_paths.resolved(target, vt.results_dir)
+    if self.context.products.is_required_data(NodePathsLocal):
+      node_paths_local = self.context.products.get_data(NodePathsLocal, init_func=NodePathsLocal)
+      # Always resolve targets if NodePathsLocal is required.
+      # This is crucial for `node-install` goal which builds against source code and relies on
+      # latest and nothing from the pants cache. The caching is done locally via the node_modules
+      # directory within source and managed by the underlying package manager.
+      invalidation_check = self._do_local_invalidation_check(targets)
+      with self.context.new_workunit(name='node-install', labels=[WorkUnitLabel.MULTITOOL]):
         for vt in invalidation_check.all_vts:
           target = vt.target
-          if not vt.valid:
-            resolver_for_target_type = self._resolver_for_target(target).global_instance()
-            resolver_for_target_type.resolve_target(self, target, vt.results_dir, node_paths)
-          node_paths.resolved(target, vt.results_dir)
+          resolver_for_target_type = self._resolver_for_target(target).global_instance()
+          results_dir = os.path.join(get_buildroot(), target.address.spec_path)
+          resolver_for_target_type.resolve_target(self, target, results_dir, node_paths_local,
+                                                  resolve_locally=True,
+                                                  force=True,
+                                                  install_optional=True,
+                                                  frozen_lockfile=False)
```

Comment:
`frozen_lockfile=False` is really fishy... is the default behaviour of `yarn install` to ignore the lockfile...?

Reply:
`--frozen-lockfile=False` is the default behavior of `yarn install`. We turn on frozen-lockfile for `.pants.d` installs because we want them to atomic. Outside of `.pants.d`, it is meaningful to produce a new lockfile such as the case of adding a new dependency.

Reply:
Ok. It makes some sense to align with Yarn's default. But this is still a pretty wacky default, IMO. It might be worth diverging from the default to be a bit more sane, but up to you.